### ISSUE TITLE
Some user-set aliases get ignored

### ISSUE THREAD

**Issue by TomSmeets** [NixOS/nixpkgs#24560]
## Issue description
The shell aliases that get set inside [nixos/modules/programs/shell.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/shell.nix#L17-L21')  will overide the user-set shellAliases.

I think that this module should use `lib.mkDefault` for each seperate alias. But I did not get that working.
#### configuration.nix
```nix
{ ... }:
{
  environment.shellAliases = {
    l = "ls -laXF --group-directories-first";
  };
}
```
#### bash
```shell
> alias
l='ls -alh'
ll='ls -l'
ls='ls --color=tty'
run-help=man
which-command=whence
```

## Technical details
* System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...)
`17.09.git.03a969d (Hummingbird)`
* Nix version: (run `nix-env --version`)
* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`)

**Comment by TomSmeets** [comment]
After exploring this issue I have a potential fix.

This fix would be to move the predefined aliases in [nixos/modules/programs/shell.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/shell.nix#L17) to the `environment.shellAliases` option as a default value.

But as soon as the user defines any alias in `environment.shellAliases` de defaults get completely ignored. Using a custom `merge` function for this option should solve this.

Another bonus of using `default` from mkOption is that it will be documented in the nixos options manual.

Please tell me what you think.

**Comment by cleverca22** [comment]
https://github.com/NixOS/nixpkgs/blob/ddb6d0962e56c8da567e89ca0e132154507616af/nixos/modules/config/shells-environment.nix#L117
i think the issue, is that the `types.attrs` type doesnt clearly define the merge policy when there are conflicting keys

**Comment by cleverca22** [comment]
just discovered that `programs.bash.shellAliases` has priority over `environment.shellAliases` and causes all of `environment.shellAliases` to be ignored

**Comment by TomSmeets** [comment]
I can confirm that this issue has been fixed by: #44441 
Closing
